File: frontend/src/lib/services/substitution.ts
import { API } from '$lib/config/api';
import { apiFetch } from '$lib/services/apiClient';

export interface SubstituteSuggestion {
	leave_application_id: string;
	teacher_id?: string | null;
	substitute_teacher_id: string;
	substitute_name: string;
	class_id: string;
	class_name?: string | null;
	subject_id?: string | null;
	subject_name?: string | null;
	day_of_week: string;
	period: number;
	score?: number | null;
	reason?: string | null;
}

export interface Substitution {
	id: string;
	institution_id?: string | null;
	leave_application_id?: string | null;
	teacher_id?: string | null;
	substitute_teacher_id?: string | null;
	class_id?: string | null;
	subject_id?: string | null;
	day_of_week?: string | null;
	period?: number | null;
	status?: string | null;
	confirmed_by?: string | null;
	created_at?: string | null;
	teacher_name?: string | null;
	substitute_name?: string | null;
	class_name?: string | null;
	subject_name?: string | null;
	leave_type?: string | null;
	leave_start?: string | null;
	leave_end?: string | null;
}

const base = `${API.baseUrl}/api/substitutions`;

export async function suggestSubstitutes(leaveApplicationId: string): Promise<SubstituteSuggestion[]> {
	return apiFetch<SubstituteSuggestion[]>(`${base}/suggest`, {
		method: 'POST',
		body: JSON.stringify({ leave_application_id: leaveApplicationId })
	});
}

export async function confirmSubstitution(data: {
	leave_application_id: string;
	teacher_id?: string;
	substitute_teacher_id: string;
	class_id: string;
	subject_id?: string;
	day_of_week: string;
	period: number;
}): Promise<Substitution> {
	return apiFetch<Substitution>(base, {
		method: 'POST',
		body: JSON.stringify(data)
	});
}

export async function getSubstitutions(status?: string): Promise<Substitution[]> {
	const qs = status ? `?status_filter=${encodeURIComponent(status)}` : '';
	return apiFetch<Substitution[]>(`${base}${qs}`);
}

File: backend/app/schemas/substitution.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel


class SubstituteSuggestionRequest(BaseModel):
    leave_application_id: UUID


class SubstituteSuggestion(BaseModel):
    leave_application_id: UUID
    teacher_id: UUID | None = None
    substitute_teacher_id: UUID
    substitute_name: str
    class_id: UUID
    class_name: str | None = None
    subject_id: UUID | None = None
    subject_name: str | None = None
    day_of_week: str
    period: int
    score: int | None = None
    reason: str | None = None


class SubstitutionCreate(BaseModel):
    leave_application_id: UUID
    teacher_id: UUID | None = None
    substitute_teacher_id: UUID
    class_id: UUID
    subject_id: UUID | None = None
    day_of_week: str
    period: int


class SubstitutionResponse(BaseModel):
    id: UUID
    institution_id: UUID | None = None
    leave_application_id: UUID | None = None
    teacher_id: UUID | None = None
    substitute_teacher_id: UUID | None = None
    class_id: UUID | None = None
    subject_id: UUID | None = None
    day_of_week: str | None = None
    period: int | None = None
    status: str | None = None
    confirmed_by: UUID | None = None
    created_at: datetime | None = None
    teacher_name: str | None = None
    substitute_name: str | None = None
    class_name: str | None = None
    subject_name: str | None = None
    leave_type: str | None = None
    leave_start: str | None = None
    leave_end: str | None = None

    model_config = {"from_attributes": True}